File: ejemplos/nerve/03-peticion-respuesta/cliente.js
const { NexusClient } = require('nerve');

class ClientePeticion {
    constructor() {
        this.cliente = new NexusClient();
        this.respuestaRecibida = null;
    }

    manejarRespuesta = (mensajeCrudo) => {
        const payload = mensajeCrudo.payload || {};
        const accion = payload.accion;

        if (accion === "respuesta_operacion") {
            this.respuestaRecibida = payload.resultado;
            console.log(`✅ ¡El servidor respondió! El resultado es: ${this.respuestaRecibida}`);
        }
    }

    async iniciar() {
        await this.cliente.connect("cliente_calculadora");
        this.cliente.listen(this.manejarRespuesta);

        console.log("Enviando petición de suma al servidor (5 + 7)...");
        const peticion = { accion: "sumar", a: 5, b: 7 };
        this.cliente.send("servidor_calculadora", peticion);

        let tiempoEspera = 0;
        while (this.respuestaRecibida === null && tiempoEspera < 5) {
            console.log("⏳ Esperando respuesta...");
            await new Promise(r => setTimeout(r, 1000));
            tiempoEspera++;
        }

        if (this.respuestaRecibida === null) {
            console.log("❌ Tiempo de espera agotado. El servidor no contestó.");
        }

        this.cliente.disconnect();
    }
}

if (require.main === module) {
    const app = new ClientePeticion();
    app.iniciar();
}

File: ejemplos/nerve/03-peticion-respuesta/cliente.go
package main

import (
	"fmt"
	"time"

	"nerve"
)

type ClientePeticion struct {
	cliente           *nerve.NexusClient
	respuestaRecibida *float64
}

func (c *ClientePeticion) manejarRespuesta(mensajeCrudo nerve.Message) {
	payload, ok := mensajeCrudo.Payload.(map[string]interface{})
	if !ok {
		return
	}

	accion, ok := payload["accion"].(string)
	if ok && accion == "respuesta_operacion" {
		if resultado, ok := payload["resultado"].(float64); ok {
			c.respuestaRecibida = &resultado
			fmt.Printf("✅ ¡El servidor respondió! El resultado es: %v\n", resultado)
		}
	}
}

func (c *ClientePeticion) iniciar() {
	c.cliente.Connect("cliente_calculadora")
	c.cliente.Listen(c.manejarRespuesta)

	fmt.Println("Enviando petición de suma al servidor (5 + 7)...")
	peticion := map[string]interface{}{
		"accion": "sumar",
		"a":      5,
		"b":      7,
	}
	c.cliente.Send("servidor_calculadora", peticion)

	tiempoEspera := 0
	for c.respuestaRecibida == nil && tiempoEspera < 5 {
		fmt.Println("⏳ Esperando respuesta...")
		time.Sleep(1 * time.Second)
		tiempoEspera++
	}

	if c.respuestaRecibida == nil {
		fmt.Println("❌ Tiempo de espera agotado. El servidor no contestó.")
	}

	c.cliente.Disconnect()
}

func main() {
	app := &ClientePeticion{
		cliente: nerve.NewNexusClient(),
	}
	app.iniciar()
}
